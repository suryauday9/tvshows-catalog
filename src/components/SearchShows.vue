<template>
    <v-container grid-list-lg v-if="showSearchbar">
        <v-layout row wrap>
            <v-flex xs12 sm12 md12 lg12>
                <v-text-field
                    prepend-inner-icon="mdi-magnify"
                    rounded
                    outlined
                    solo
                    clearable
                    color="blue darken-2"
                    dark
                    autofocus
                    placeholder="Search Shows"
                    v-on:keyup="searchShows" 
                    v-model="searchText"
                    @click:clear="clearSearch"
                />
            </v-flex>
        </v-layout>
    </v-container>
</template>

<script>
export default {
    name: 'SearchShows',
    data() {
        return {
            searchText: '',
        }
    },
    computed: {
        showSearchbar() {
            return this.$store.state.showSearchbar;
        }
    },
    watch: {
        'showSearchbar': function (value) {
            if(!value) this.clearSearch();
        }
    },
    methods: {
        searchShows() {
            if(this.searchText && this.searchText.length >= 3) {
                this.$store.dispatch('getShowsBySearch', this.searchText);
            } else {
                this.$store.dispatch('getShowslist');
            }
        },
        clearSearch() {
            this.searchText = "";
            this.searchShows();
        }
    }
};
</script>

<style scoped>
    .selected-text {
        font-size: 16px !important;
        font-weight: 500;
        display: block;
        margin: 6px 9px;
        text-align: center;
    }
</style>
