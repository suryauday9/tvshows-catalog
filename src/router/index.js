import Vue from "vue";
import VueRouter from "vue-router";

Vue.use(VueRouter);

const routes = [
  {
    path: "/",
    name: 'ShowsCatalog',
    component: () => import('../views/ShowsCatalog.vue')
  },
  {
    path: '/show',
    name: 'show',
    component: () =>
      import('../views/ShowDetails.vue'),
    props: true,
  },
  {
    path: '/404',
    alias: '*',
    name: 'notFound',
    component: () =>
      import('../views/Page404.vue'),
  },
];

const router = new VueRouter({
  mode: "history",
  base: process.env.BASE_URL,
  routes
});

export default router;
