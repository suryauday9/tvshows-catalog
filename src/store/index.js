import Vue from "vue";
import Vuex from "vuex";

import {
getAllShows,
getShowDetails,
getShowThumbnails,
getShowsBySearch
} from "../services/service";

Vue.use(Vuex);

const state = {
  shows: [],
  genres: [],
  showDetails: [],
  showThumbnails: [],
  showSearchbar: false
};

export const sortByRating = (response) => {
  return response.sort((prevValue, nextValue) =>
  prevValue.rating.average < nextValue.rating.average ? 1 : -1);
};

export const getGenres = (response) => {
  return response
  .reduce((result, show) => result.concat(show.genres), [])
  .filter((genre, index, self) => self.indexOf(genre) === index);
};

export const actions = {
  async getShowslist({ commit }) {
    getAllShows().then((response) => {
      commit('setShowslist', sortByRating(response));
      commit('setGenreslist', getGenres(response));
    })
    .catch((error) => {
      const errorNotification = {
        type: 'error',
        message: `There was a problem getting all shows: ${error.message}`,
      };
      console.log(errorNotification);
    });
  },
  async getShowDetails({ commit }, showId) {
    getShowDetails(showId).then((response) => {
      commit('setShowDetails', response);
    })
    .catch((error) => {
      const errorNotification = {
        type: 'error',
        message: `There was a problem getting show details: ${error.message}`,
      };
      console.log(errorNotification);
    });
  },
  async getShowThumbnails({ commit }, showId) {
    await getShowThumbnails(showId).then((response) => {
      commit('setShowThumbnails', response);
    })
    .catch((error) => {
      const errorNotification = {
        type: 'error',
        message: `There was a problem getting show thumbnails: ${error.message}`,
      };
      console.log(errorNotification);
    });
  },
  async getShowsBySearch({ commit }, searchTerm) {
    await getShowsBySearch(searchTerm).then((response) => {
      let resultSet = response.map(result => (result.show));
      commit('setShowslist', resultSet);
      commit('setGenreslist', getGenres(resultSet));
    })
    .catch((error) => {
      const errorNotification = {
        type: 'error',
        message: `There was a problem in show search: ${error.message}`,
      };
      console.log(errorNotification);
    });
  },
  getSearchbarStatus ({ commit }) {
    commit('setSearchbarStatus', !this.state.showSearchbar);
  }
};

export const mutations = {
  setShowslist(state, payload) {
    state.shows = payload;
  },
  setGenreslist(state, payload) {
      state.genres = payload;
  },
  setShowDetails(state, payload) {
    state.showDetails = payload;
  },
  setShowThumbnails(state, payload) {
    state.showThumbnails = payload;
  },
  setSearchbarStatus(state, payload) {
    state.showSearchbar = payload;
  }
};

export default new Vuex.Store({
  state: state,
  actions: actions,
  mutations: mutations,
  modules: {}
});
