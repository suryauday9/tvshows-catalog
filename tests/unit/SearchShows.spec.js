import { mount, shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import Vuetify from 'vuetify';
import VueRouter from 'vue-router';

import { initializeApp } from './initializeApp';
import SearchShows from '../../src/components/SearchShows.vue';
import { showsData } from './response';
import { routes } from '../../src/router';

const vuetify = new Vuetify();
const localVue = initializeApp(createLocalVue());

describe('PageHeader component', () => {  
    let wrapper;
    const router = new VueRouter(routes);
    
    const store = new Vuex.Store({
        state: {
            shows: showsData,
            genres: [],
            showDetails: [],
            showThumbnails: [],
            showSearchbar: false
        },
        actions: {
            getShowsBySearch: jest.fn(),
            getShowslist: jest.fn()
        }
    });

    const mountComponent = (options) => {
        return shallowMount(SearchShows, {
            localVue,
            router,
            store,
            data() {
                return {
                    searchText: 'game'
                }
            },
            vuetify,
            ...options,
        })
    }

    beforeEach(() => {
        wrapper = mountComponent();
    })

    afterEach(() => {
        wrapper.destroy();
    })

    it('renders component correctly', () => {
        // see if component renders
        expect(wrapper.element).toMatchSnapshot();
    });

    it('should be a Vue instance', () => {
        expect(wrapper.isVueInstance).toBeTruthy();
    });

    it('rendered component data should match with store', () => {
        // see if data loads
        expect(wrapper.vm.showSearchbar).toBe(store.state.showSearchbar);
    });

    it('should dispatch getShowslist method', () => {
        const searchWrapper = mount(SearchShows, {
            localVue,
            vuetify,
            store,
            router,
            data() {
                return {
                    searchText: ''
                }
            },
        });
        const actn = searchWrapper.vm.searchShows();
        expect(actn).toBeUndefined();
        expect(searchWrapper.vm.$nextTick()).toStrictEqual(store.dispatch('getShowslist'));
    });

    it('should dispatch getShowsBySearch method', () => {
        const searchWrapper = mount(SearchShows, {
            localVue,
            vuetify,
            store,
            router,
            data() {
                return {
                    searchText: 'game'
                }
            },
        });
        const actn = searchWrapper.vm.searchShows();
        expect(actn).toBeUndefined();
        expect(searchWrapper.vm.$nextTick()).toStrictEqual(store.dispatch('getShowsBySearch'));
    });

    it('should trigger clearSearch method', () => {
        const searchWrapper = mount(SearchShows, {
            localVue,
            vuetify,
            store,
            router,
            data() {
                return {
                    searchText: 'game'
                }
            },
        });
        const actn = searchWrapper.vm.clearSearch();
        expect(actn).toBeUndefined();
        expect(searchWrapper.vm.$nextTick()).toStrictEqual(store.dispatch('getShowsBySearch'));
    });

    it('should trigger the watch service for showSearchbar', () => {
        expect(wrapper.vm.$options.watch.showSearchbar.call(wrapper.vm)).toBeUndefined();
        expect(wrapper.vm.$options.watch.showSearchbar.call(wrapper.vm, true)).toBeUndefined();
    });

});