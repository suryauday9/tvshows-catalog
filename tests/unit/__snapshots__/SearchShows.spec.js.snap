// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageHeader component renders component correctly 1`] = `<!---->`;
